eZ/Publish/Core/Persistence/Cache/InMemory/InMemoryCache.php

```diff
@@ -0,0 +1,203 @@
+        $this->enabled = $enabled;
+    }
+
+    /**
+     * Returns a cache objects.
+     *
+     * @param string $key Primary or secondary index to look for cache on.
+     *
+     * @return object|null Object if found, null if not.
+     */
+    public function get(string $key)
+    {
+        if ($this->enabled === false) {
+            return null;
+        }
+
+        // Check for global expiry change, if the case clear cache
+        if ($this->cacheExpiryTime !== self::$globalCacheExpiry) {
+            $this->clear();
+            $this->cacheExpiryTime = self::$globalCacheExpiry;
+
+            return null;
+        }
+
+        $index = $this->cacheIndex[$key] ?? $key;
+        if (!isset($this->cache[$index]) || $this->cacheTime[$index] + $this->ttl < microtime(true)) {
+            return null;
+        }
+
+        return $this->cache[$index];
+    }
+
+    /**
+     * Set object in in-memory cache.
+     *
+     * Should only set Cache hits here!
+     *
+     * @param object[] $objects
+     * @param callable $objectIndexes Return array of indexes per object (first argument), must return at least 1 primary index
+     * @param string|null $listIndex Optional index for list of items
+     */
+    public function setMulti(array $objects, callable $objectIndexes, string $listIndex = null): void
+    {
+        // If objects accounts for more then 20% of our limit, assume it's bulk content load and skip saving in-memory
+        if ($this->enabled === false || \count($objects) >= $this->limit / 5) {
+            return;
+        }
+
+        // check if we will reach limit by adding these objects, if so remove old cache
+        if (\count($this->cache) + \count($objects) >= $this->limit) {
+            $this->vacuum();
+        }
+
+        $time = microtime(true);
+
+        // if set add objects to cache on list index (typically a "all" key)
+        if ($listIndex) {
+            $this->cache[$listIndex] = $objects;
+            $this->cacheTime[$listIndex] = $time;
+        }
+
+        foreach ($objects as $object) {
+            // Skip if there are no indexes
+            if (!$indexes = $objectIndexes($object)) {
+                continue;
+            }
+
+            $key = array_shift($indexes);
+            $this->cache[$key] = $object;
+            $this->cacheTime[$key] = $time;
+
+            foreach ($indexes as $index) {
+                $this->cacheIndex[$index] = $key;
+            }
+        }
+    }
+
+    /**
+     * Removes multiple in-memory cache from the pool.
+     *
+     * @param string[] $keys An array of keys that should be removed from the pool.
+     */
+    public function deleteMulti(array $keys): void
+    {
+        if ($this->enabled === false) {
+            return;
+        }
+
+        foreach ($keys as $key) {
+            if ($index = $this->cacheIndex[$key] ?? null) {
+                unset($this->cacheIndex[$key], $this->cache[$index], $this->cacheTime[$index]);
```

Comment:
q: what if a key is in both $this->cacheIndex and $this->cache ?

Reply:
we can indeed clear for that to as there is no validation to prevent it atm

That said, be aware this has a big fat `@internal` on it, so this is not for you @gggeek, not yet.